### PR TITLE
Use shared informers for secrets, services, and service accounts

### PR DESCRIPTION
**Pull request by smarterclayton**
Consolidate existing informers into shared informers.

The upstream token controller may need mutationcache

[test]

### CODE REVIEW

File: pkg/security/controller/namespace_security_allocation_controller.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		return label
+	}
+}
+
+// Next processes a changed namespace and tries to allocate a uid range for it.  If it is
+// successful, an mcs label corresponding to the relative position of the range is also
+// set.
+func (c *NamespaceSecurityDefaultsController) allocate(ns *kapi.Namespace) error {
+	tx := &tx{}
+	defer tx.Rollback()
```
Comment (kargakis) [inline pkg/security/controller/namespace_security_allocation_controller.go]: These defers do not need to run before uid allocation.

**Comment by smarterclayton** [comment]
But they're free and can't accidentally be broken here.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Two log rate limit(?) related flakes and one compile [test]

**Comment by smarterclayton** [comment]
@stevekuznetsov wat:
https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_install/1848/consoleFull#-157296560358b6e51eb7608a5981914356

**Comment by stevekuznetsov** [comment]
@smarterclayton #14122 re[test]

**Comment by smarterclayton** [comment]
[test] super happy that this is mostly going green

**Comment by smarterclayton** [comment]
[test] might be flakes

**Comment by smarterclayton** [comment]
[test] startup flake

**Comment by smarterclayton** [comment]
more [test] to counter flake.

**Comment by smarterclayton** [comment]
@deads2k token_controller does:

  e.secrets.Add(createdToken)
​
Can I simply port EtcdMutationCache upstream and replace this?

**Comment by deads2k** [comment]
> Can I simply port EtcdMutationCache upstream and replace this?

Yes

**Comment by smarterclayton** [comment]
Done

**Comment by smarterclayton** [comment]
[test] various weirdnesses

**Comment by smarterclayton** [comment]
@stevekuznetsov i can't figure out why test-go is reporting a failure.

**Comment by stevekuznetsov** [comment]
```
[WARNING] While the jUnit report found no failed tests, the `go test` process failed.
[WARNING] This usually means that the unit test suite failed to compile.
[WARNING] `go test` detected data races.
[WARNING] Details can be found in the full output file at lines 29288 29338 29390 29445 29495 29547 29603 29653 29705 29761 29811 29863 29915 29967 30037 30087 30139 30191 30257 30307 30359.
```

**Comment by stevekuznetsov** [comment]
I opened an issue upstream about this a while ago -- a data race fails the process but doesn't fail the test IIRC.

File: pkg/cmd/server/origin/run_components.go [hunk 2]
```diff
@@ -160,30 +163,52 @@ func (c *MasterConfig) RunServiceAccountTokensController(cm *cmapp.CMServer) {
...
-		DockercfgController:  dockercfgController,
-		DockerURLsIntialized: dockerURLsIntialized,
+		RegistryNamespace:     "default",
+		RegistryServiceName:   "docker-registry",
+		DockercfgController:   dockercfgController,
+		DockerURLsInitialized: dockerURLsInitialized,
 	}
-	go serviceaccountcontrollers.NewDockerRegistryServiceController(c.KubeClientsetInternal(), dockerRegistryControllerOptions).Run(10, make(chan struct{}))
+	go serviceaccountcontrollers.NewDockerRegistryServiceController(
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().Secrets().Informer(),
```
Comment (mfojtik) [inline pkg/cmd/server/origin/run_components.go]: can you also add service shared informer?

File: pkg/cmd/server/start/start_master.go [hunk 3]
```diff
@@ -602,6 +602,9 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 	// Start these first, because they provide credentials for other controllers' clients
 	oc.RunServiceAccountsController()
 	oc.RunServiceAccountTokensController(controllerManagerOptions)
+	// Must start a few informers now
+	oc.Informers.KubernetesInformers().Start(utilwait.NeverStop)
```
Comment (mfojtik) [inline pkg/cmd/server/start/start_master.go]: why do we have to do this?

File: pkg/service/controller/servingcert/secret_creating_controller.go [hunk 4]
```diff
@@ -87,17 +83,8 @@ func NewServiceServingCertController(serviceClient kcoreclient.ServicesGetter, s
...
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return sc.serviceClient.Services(metav1.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return sc.serviceClient.Services(metav1.NamespaceAll).Watch(options)
-			},
-		},
-		&kapi.Service{},
-		resyncInterval,
+	sc.serviceCache = services.GetStore()
```
Comment (mfojtik) [inline pkg/service/controller/servingcert/secret_creating_controller.go]: why not do this in `&ServiceServingCertController{..}` ?

File: pkg/service/controller/servingcert/secret_updating_controller.go [hunk 5]
```diff
@@ -62,38 +57,22 @@ func NewServiceServingCertUpdateController(serviceClient kcoreclient.ServicesGet
...
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return serviceClient.Services(metav1.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return serviceClient.Services(metav1.NamespaceAll).Watch(options)
-			},
-		},
-		&kapi.Service{},
-		resyncInterval,
+	sc.serviceCache = services.GetStore()
```
Comment (mfojtik) [inline pkg/service/controller/servingcert/secret_updating_controller.go]: why not GetIndexer() ?

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go [hunk 6]
```diff
@@ -53,32 +50,23 @@ type DockercfgControllerOptions struct {
 	// If zero, re-list will be delayed as long as possible
 	Resync time.Duration
 
-	// DockerURLsIntialized is used to send a signal to this controller that it has the correct set of docker urls
+	// DockerURLsInitialized is used to send a signal to this controller that it has the correct set of docker urls
 	// This is normally signaled from the DockerRegistryServiceController which watches for updates to the internal
 	// docker registry service.
-	DockerURLsIntialized chan struct{}
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: nice catch :)

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go [hunk 7]
```diff
@@ -53,32 +50,23 @@ type DockercfgControllerOptions struct {
...
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return e.client.Core().ServiceAccounts(api.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return e.client.Core().ServiceAccounts(api.NamespaceAll).Watch(options)
-			},
-		},
-		&api.ServiceAccount{},
-		options.Resync,
+	serviceAccountCache := serviceAccounts.GetStore()
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: @smarterclayton same here... is there a rule of thumb when to use GetStore() vs. GetIndexer() ?

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 8]
```diff
@@ -269,9 +250,18 @@ func (e *DockerRegistryServiceController) syncRegistryLocationChange(key string)
 	// we've changed the docker registry URL.  Add items to the work queue for all known secrets
 	// new secrets will already get the updated value.
 	for _, obj := range e.secretCache.List() {
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: @smarterclayton i think this needs lister?

**Comment by deads2k** [comment]
@smarterclayton I needed the mutation cache in https://github.com/kubernetes/kubernetes/pull/45838 .  I've made some tweaks to get it re-useable for generic controllers.

File: pkg/cmd/server/origin/run_components.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/run_components.go]: It's not needed, we only have a single item we are watching, and it's less efficient to filter the entire informer set for one items

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: Same object, use the one that fits the needed API.

**Comment by deads2k** [comment]
List of controller dispositions

lgtm
 1. pkg/project/auth/cache.go - lgtm

have comments

File: pkg/project/cache/cache.go [hunk 9]
```diff
@@ -18,16 +17,24 @@ import (
...
 		Client:              client,
+		Store:               namespaces.GetIndexer(),
+		Controller:          namespaces.GetController(),
 		DefaultNodeSelector: defaultNodeSelector,
 	}
 }
 
 type ProjectCache struct {
 	Client              kcoreclient.NamespaceInterface
 	Store               cache.Indexer
```
Comment (deads2k) [inline pkg/project/cache/cache.go]: this exposure is nasty.  Doesn't have to be fixed here.

File: pkg/project/cache/cache.go [hunk 10]
```diff
@@ -18,16 +17,24 @@ import (
...
+	Controller          cache.Controller
```
Comment (deads2k) [inline pkg/project/cache/cache.go]: If you're only using hasSynced, use a hasSynced method.

File: pkg/project/controller/project_finalizer_controller.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+func NewProjectFinalizerController(namespaces cache.SharedIndexInformer, client internalclientset.Interface) *ProjectFinalizerController {
+	c := &ProjectFinalizerController{
+		client:     client,
+		controller: namespaces.GetController(),
+		cache:      namespaces.GetStore(),
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+	namespaces.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
```
Comment (deads2k) [inline pkg/project/controller/project_finalizer_controller.go]: TODO, I think we want to generalize my naiveResourceEventHandler: https://github.com/openshift/origin/blob/master/pkg/authorization/controller/authorizationsync/generic.go#L71 .  It seems pretty common.

File: pkg/project/controller/project_finalizer_controller.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+	c.syncHandler = c.syncNamespace
+	return c
+}
+
+// Run starts the workers for this controller.
+func (c *ProjectFinalizerController) Run(stopCh <-chan struct{}, workers int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
```
Comment (deads2k) [inline pkg/project/controller/project_finalizer_controller.go]: glogs for start and stop please.

File: pkg/security/controller/namespace_security_allocation_controller.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+	if !ok {
+		return
+	}
+	c.queue.Add(ns.Name)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *NamespaceSecurityDefaultsController) worker() {
+	for {
```
Comment (deads2k) [inline pkg/security/controller/namespace_security_allocation_controller.go]: we had a way of making this one liner, didn't we?

File: pkg/security/controller/namespace_security_allocation_controller.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+	c.syncHandler = c.syncNamespace
+	return c
+}
+
+// Run starts the workers for this controller.
+func (c *NamespaceSecurityDefaultsController) Run(stopCh <-chan struct{}, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the stores to fill
```
Comment (deads2k) [inline pkg/security/controller/namespace_security_allocation_controller.go]: glog for start and stop

File: pkg/security/controller/namespace_security_allocation_controller.go [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+	return c.allocate(item.(*kapi.Namespace))
+}
+
+type MCSAllocationFunc func(uid.Block) *mcs.Label
+
+// DefaultMCSAllocation returns a label from the MCS range that matches the offset
+// within the overall range. blockSize must be a positive integer representing the
+// number of labels to jump past in the category space (if 1, range == label, if 2
+// each range will have two labels).
+func DefaultMCSAllocation(from *uid.Range, to *mcs.Range, blockSize int) MCSAllocationFunc {
```
Comment (deads2k) [inline pkg/security/controller/namespace_security_allocation_controller.go]: straight move?

File: pkg/service/controller/ingressip/service_ingressip_controller.go [hunk 16]
```diff
@@ -167,30 +166,22 @@ func (ic *IngressIPController) enqueueChange(new interface{}, old interface{}) {
...
-		if !ic.work() {
-			break
-		}
+	glog.V(5).Infof("Starting worker")
+	for ic.work() {
 	}
 
 	glog.V(5).Infof("Shutting down ingress ip controller")
-	ic.queue.ShutDown()
 }
```
Comment (deads2k) [inline pkg/service/controller/ingressip/service_ingressip_controller.go]: I'm not seeing the block on stopCh

File: pkg/service/controller/ingressip/service_ingressip_controller.go [hunk 17]
```diff
@@ -167,30 +166,22 @@ func (ic *IngressIPController) enqueueChange(new interface{}, old interface{}) {
...
 		return
 	}
 
-	glog.V(5).Infof("Starting normal worker")
-	for {
-		if !ic.work() {
-			break
-		}
+	glog.V(5).Infof("Starting worker")
+	for ic.work() {
```
Comment (deads2k) [inline pkg/service/controller/ingressip/service_ingressip_controller.go]: finish our normal structure? I can't see at a glance here if this handles panics properly with a backoff and retry.

File: pkg/service/controller/servingcert/secret_creating_controller.go [hunk 18]
```diff
@@ -75,7 +71,7 @@ type ServiceServingCertController struct {
 
 // NewServiceServingCertController creates a new ServiceServingCertController.
 // TODO this should accept a shared informer
-func NewServiceServingCertController(serviceClient kcoreclient.ServicesGetter, secretClient kcoreclient.SecretsGetter, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertController {
+func NewServiceServingCertController(services, secrets cache.SharedIndexInformer, serviceClient kcoreclient.ServicesGetter, secretClient kcoreclient.SecretsGetter, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertController {
```
Comment (deads2k) [inline pkg/service/controller/servingcert/secret_creating_controller.go]: please pass particular informers.

File: pkg/service/controller/servingcert/secret_creating_controller.go [hunk 19]
```diff
@@ -150,8 +130,6 @@ func (sc *ServiceServingCertController) Run(workers int, stopCh <-chan struct{})
 	}
 
 	<-stopCh
-	glog.Infof("Shutting down service signing cert controller")
```
Comment (deads2k) [inline pkg/service/controller/servingcert/secret_creating_controller.go]: we like the paired glogs

File: pkg/service/controller/servingcert/secret_updating_controller.go [hunk 20]
```diff
@@ -50,7 +45,7 @@ type ServiceServingCertUpdateController struct {
 
 // NewServiceServingCertUpdateController creates a new ServiceServingCertUpdateController.
 // TODO this should accept a shared informer
-func NewServiceServingCertUpdateController(serviceClient kcoreclient.ServicesGetter, secretClient kcoreclient.SecretsGetter, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertUpdateController {
+func NewServiceServingCertUpdateController(services, secrets cache.SharedIndexInformer, secretClient kcoreclient.SecretsGetter, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertUpdateController {
```
Comment (deads2k) [inline pkg/service/controller/servingcert/secret_updating_controller.go]: the individual informers should be typed.

File: pkg/service/controller/servingcert/secret_updating_controller.go [hunk 21]
```diff
@@ -103,52 +82,22 @@ func NewServiceServingCertUpdateController(serviceClient kcoreclient.ServicesGet
 // Run begins watching and syncing.
 func (sc *ServiceServingCertUpdateController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
-	defer glog.Infof("Shutting down service signing cert update controller")
 	defer sc.queue.ShutDown()
 
-	glog.Infof("starting service signing cert update controller")
```
Comment (deads2k) [inline pkg/service/controller/servingcert/secret_updating_controller.go]: we want these.  At least until we have reasonable health checks for them.

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go [hunk 22]
```diff
@@ -53,32 +50,23 @@ type DockercfgControllerOptions struct {
...
+	// DockerURLsInitialized is used to send a signal to this controller that it has the correct set of docker urls
 	// This is normally signaled from the DockerRegistryServiceController which watches for updates to the internal
 	// docker registry service.
-	DockerURLsIntialized chan struct{}
+	DockerURLsInitialized chan struct{}
 }
 
 // NewDockercfgController returns a new *DockercfgController.
-func NewDockercfgController(cl kclientset.Interface, options DockercfgControllerOptions) *DockercfgController {
+func NewDockercfgController(serviceAccounts, secrets cache.SharedIndexInformer, cl kclientset.Interface, options DockercfgControllerOptions) *DockercfgController {
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: specific input types

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go [hunk 23]
```diff
@@ -92,30 +80,31 @@ func NewDockercfgController(cl kclientset.Interface, options DockercfgController
...
-	tokenSecretSelector := fields.OneTermEqualSelector(api.SecretTypeField, string(api.SecretTypeServiceAccountToken))
-	e.secretCache, e.secretController = cache.NewInformer(
-		&cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				options.FieldSelector = tokenSecretSelector.String()
-				return e.client.Core().Secrets(api.NamespaceAll).List(options)
+	e.secretCache = secrets.GetIndexer()
+	e.secretController = secrets.GetController()
+	secrets.AddEventHandlerWithResyncPeriod(
+		cache.FilteringResourceEventHandler{
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: I like this.  Pre-existing or did you create it in this pull?

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go [hunk 24]
```diff
@@ -216,29 +206,27 @@ func (e *DockercfgController) Run(workers int, stopCh <-chan struct{}) {
 	case <-stopCh:
 		return
 	}
-	glog.Infof("Dockercfg secret controller initialized, starting.")
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: preserve

File: pkg/serviceaccounts/controllers/deleted_dockercfg_secrets.go [hunk 25]
```diff
@@ -30,28 +27,28 @@ type DockercfgDeletedControllerOptions struct {
 }
 
 // NewDockercfgDeletedController returns a new *DockercfgDeletedController.
-func NewDockercfgDeletedController(cl kclientset.Interface, options DockercfgDeletedControllerOptions) *DockercfgDeletedController {
+func NewDockercfgDeletedController(secrets cache.SharedIndexInformer, cl kclientset.Interface, options DockercfgDeletedControllerOptions) *DockercfgDeletedController {
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/deleted_dockercfg_secrets.go]: specific types

File: pkg/serviceaccounts/controllers/deleted_token_secrets.go [hunk 26]
```diff
@@ -24,28 +23,28 @@ type DockercfgTokenDeletedControllerOptions struct {
...
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				opts := metav1.ListOptions{FieldSelector: dockercfgSelector.String()}
-				return e.client.Core().Secrets(api.NamespaceAll).List(opts)
+	e.secretController = secrets.GetController()
+	secrets.AddEventHandlerWithResyncPeriod(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *api.Secret:
+					return t.Type == api.SecretTypeServiceAccountToken
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/deleted_token_secrets.go]: Looks like we need a `SecretFilterFunc(secretType)`

File: pkg/serviceaccounts/controllers/deleted_token_secrets.go [hunk 27]
```diff
@@ -24,28 +23,28 @@ type DockercfgTokenDeletedControllerOptions struct {
 }
 
 // NewDockercfgTokenDeletedController returns a new *DockercfgTokenDeletedController.
-func NewDockercfgTokenDeletedController(cl kclientset.Interface, options DockercfgTokenDeletedControllerOptions) *DockercfgTokenDeletedController {
+func NewDockercfgTokenDeletedController(secrets cache.SharedIndexInformer, cl kclientset.Interface, options DockercfgTokenDeletedControllerOptions) *DockercfgTokenDeletedController {
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/deleted_token_secrets.go]: specific types

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 28]
```diff
@@ -35,22 +35,23 @@ type DockerRegistryServiceControllerOptions struct {
...
 
-	// DockerURLsIntialized is used to send a signal to the DockercfgController that it has the correct set of docker urls
-	DockerURLsIntialized chan struct{}
+	// DockerURLsInitialized is used to send a signal to the DockercfgController that it has the correct set of docker urls
+	DockerURLsInitialized chan struct{}
 }
 
 // NewDockerRegistryServiceController returns a new *DockerRegistryServiceController.
-func NewDockerRegistryServiceController(cl kclientset.Interface, options DockerRegistryServiceControllerOptions) *DockerRegistryServiceController {
+func NewDockerRegistryServiceController(secrets cache.SharedIndexInformer, cl kclientset.Interface, options DockerRegistryServiceControllerOptions) *DockerRegistryServiceController {
```
Comment (deads2k) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: specific type

File: test/cmd/policy.sh [hunk 29]
```diff
@@ -107,32 +107,32 @@ os::cmd::expect_failure_and_text 'oc policy scc-review -f ${OS_ROOT}/test/testda
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/job.yaml -o=jsonpath={.status.AllowedBy.name}' 'anyuid'
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/redis-slave.yaml -o=jsonpath={.status.AllowedBy.name}' 'anyuid'
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/nginx_pod.yaml -o=jsonpath={.status.AllowedBy.name}' 'privileged'
-os::cmd::expect_success "oc login -u bob -p bob"
+os::cmd::expect_success "oc login -u bob -p policy-missing"
```
Comment (deads2k) [inline test/cmd/policy.sh]: why change the ignored password?

File: test/cmd/policy.sh [hunk 30]
```diff
@@ -107,32 +107,32 @@ os::cmd::expect_failure_and_text 'oc policy scc-review -f ${OS_ROOT}/test/testda
...
+os::cmd::expect_success_and_text 'oc policy scc-subject-review -u bob -f ${OS_ROOT}/test/testdata/job.yaml -n policy-second --no-headers=true' 'Job/hello   <none>'
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -z default -f ${OS_ROOT}/test/testdata/job.yaml' ''
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -z default -g system:authenticated -f ${OS_ROOT}/test/testdata/job.yaml' 'restricted'
 os::cmd::expect_failure_and_text 'oc policy scc-subject-review -u alice -z default -g system:authenticated -f ${OS_ROOT}/test/testdata/job.yaml' 'error: --user and --serviceaccount are mutually exclusive'
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -z system:serviceaccount:alice:default -g system:authenticated -f ${OS_ROOT}/test/testdata/job.yaml' 'restricted'
 os::cmd::expect_success_and_text 'oc policy scc-subject-review -u alice -g system:authenticated -f ${OS_ROOT}/test/testdata/job.yaml' 'restricted'
 os::cmd::expect_success 'oc create -f ${OS_ROOT}/test/testdata/scc_lax.yaml'
-os::cmd::expect_success "oc login -u bob -p bob"
+os::cmd::expect_success "oc login -u bob -p policy-second"
```
Comment (deads2k) [inline test/cmd/policy.sh]: why change the ignored password

**Comment by deads2k** [comment]
You already have separate commits.  Can we start splitting off controllers for separate review/merge?

**Comment by smarterclayton** [comment]
Yes straight move, but the informer changed the function call

**Comment by smarterclayton** [comment]
Part of this pull

**Comment by smarterclayton** [comment]
Sigh, though -p was project.

**Comment by smarterclayton** [comment]
Yes we can split, I wanted to see everything passing stably.  I'll create upstreams first for the core code.

File: pkg/security/controller/namespace_security_allocation_controller.go (hunk 15 above)
Comment (smarterclayton) [inline pkg/security/controller/namespace_security_allocation_controller.go]: Yes straight move, then the function change.

File: pkg/service/controller/ingressip/service_ingressip_controller.go (hunk 17 above)
Comment (smarterclayton) [inline pkg/service/controller/ingressip/service_ingressip_controller.go]: Would prefer not to change the controller too much.  It does use rate limited queues, but it doesn't handle panics.  However, panics that hold locks are still going to completely break the underlying structure, so I'm not sure we should be recovering anything here without more thought.  I will queue up an issue for someone to come through and improve it.

File: pkg/serviceaccounts/controllers/create_dockercfg_secrets.go (hunk 23 above)
Comment (smarterclayton) [inline pkg/serviceaccounts/controllers/create_dockercfg_secrets.go]: net new

**Comment by smarterclayton** [comment]
Spawned kubernetes/kubernetes#45933

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Going to split into upstream stuff and downstream stuff.

**Comment by smarterclayton** [comment]
verbal lgtm from david, putting this in the queue to close the p0 memory regression in 3.6

[merge][severity: blocker]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test] boltdb panic

**Pull request merged**